File: c/tests/test_affect_vad.c
/*
 * test_affect_vad.c — AffectVad derivation tests.
 *
 * Vectors and expected values mirror fixtures/affect_vad_derivation.json
 * with epsilon 1e-5. All language ports must produce byte-identical math.
 *
 * Returns 0 on success; asserts on first failure.
 */

#include <assert.h>
#include <math.h>
#include <stdio.h>

#include "circle_ai/circle_ai.h"

#define EPSILON 1e-5f

static void check(const char *id, const char *axis, float got, float expected) {
    float diff = fabsf(got - expected);
    if (diff > EPSILON) {
        fprintf(stderr, "FAIL [%s.%s]: got %.8f, expected %.8f (diff %.8f)\n",
                id, axis, (double)got, (double)expected, (double)diff);
        assert(0);
    }
}

static void run_case(
    const char *id,
    float curiosity, float engagement, float uncertainty,
    float rapport,   float energy,
    float v_exp,     float a_exp,      float d_exp
) {
    ca_affect_state_t s = ca_affect_state_default();
    s.curiosity   = curiosity;
    s.engagement  = engagement;
    s.uncertainty = uncertainty;
    s.rapport     = rapport;
    s.energy      = energy;

    ca_affect_vad_t vad;
    ca_affect_vad_from(&s, &vad);

    check(id, "valence",   vad.valence,   v_exp);
    check(id, "arousal",   vad.arousal,   a_exp);
    check(id, "dominance", vad.dominance, d_exp);
}

int main(void) {
    /* default_state */
    run_case("default",
        0.5f, 0.5f, 0.2f, 0.0f, 0.5f,
        0.43333333f, 0.425f, 0.65f);

    /* all_max */
    run_case("all_max",
        1.0f, 1.0f, 0.0f, 1.0f, 1.0f,
        1.0f, 0.75f, 1.0f);

    /* all_min_high_uncertainty */
    run_case("all_min",
        0.0f, 0.0f, 1.0f, 0.0f, 0.0f,
        0.0f, 0.25f, 0.0f);

    /* high_engagement_warm */
    run_case("warm",
        0.6f, 0.9f, 0.1f, 0.8f, 0.7f,
        0.86666667f, 0.525f, 0.9f);

    /* stressed_low_energy */
    run_case("stressed",
        0.3f, 0.2f, 0.8f, 0.0f, 0.2f,
        0.13333333f, 0.375f, 0.2f);

    /* energetic_curious */
    run_case("energetic",
        0.9f, 0.6f, 0.3f, 0.4f, 0.9f,
        0.56666667f, 0.75f, 0.65f);

    /* NULL-safety — both pointer args required, no-op on NULL */
    ca_affect_vad_t vad;
    vad.valence = vad.arousal = vad.dominance = -1.0f;
    ca_affect_vad_from(NULL, &vad);
    assert(vad.valence   == -1.0f);
    assert(vad.arousal   == -1.0f);
    assert(vad.dominance == -1.0f);

    ca_affect_state_t s = ca_affect_state_default();
    ca_affect_vad_from(&s, NULL); /* must not crash */

    printf("All affect VAD tests passed.\n");
    return 0;
}

File: python/tests/test_affect_vad.py
"""test_affect_vad.py

Validates AffectVad derivation against all vectors in
fixtures/affect_vad_derivation.json. Math must be byte-identical across
language ports — see the fixture's $comment for the contract.
"""
from __future__ import annotations

import json
import math
import pathlib

import pytest

from circle_ai.memory.affect_state import AffectState, AffectVad

FIXTURES_DIR = pathlib.Path(__file__).parent.parent.parent / "fixtures"


def _load_fixture() -> dict:
    with open(FIXTURES_DIR / "affect_vad_derivation.json", encoding="utf-8") as f:
        return json.load(f)


FIXTURE = _load_fixture()
EPSILON = float(FIXTURE["epsilon"])
VECTORS = FIXTURE["vectors"]


def _make_state(inp: dict) -> AffectState:
    state = AffectState()
    state.curiosity   = float(inp["curiosity"])
    state.engagement  = float(inp["engagement"])
    state.uncertainty = float(inp["uncertainty"])
    state.rapport     = float(inp["rapport"])
    state.energy      = float(inp["energy"])
    return state


@pytest.mark.parametrize("vector", VECTORS, ids=[v["id"] for v in VECTORS])
def test_affect_vad_vector(vector: dict) -> None:
    """from_state() must match the canonical derivation for every fixture vector."""
    state = _make_state(vector["input"])
    vad = AffectVad.from_state(state)
    expected = vector["expected"]

    assert math.isclose(vad.valence, float(expected["valence"]), abs_tol=EPSILON), (
        f"[{vector['id']}] valence mismatch: "
        f"got {vad.valence}, expected {expected['valence']}"
    )
    assert math.isclose(vad.arousal, float(expected["arousal"]), abs_tol=EPSILON), (
        f"[{vector['id']}] arousal mismatch: "
        f"got {vad.arousal}, expected {expected['arousal']}"
    )
    assert math.isclose(vad.dominance, float(expected["dominance"]), abs_tol=EPSILON), (
        f"[{vector['id']}] dominance mismatch: "
        f"got {vad.dominance}, expected {expected['dominance']}"
    )


def test_to_vad_returns_same_as_from_state() -> None:
    """AffectState.to_vad() must be equivalent to AffectVad.from_state(state)."""
    state = AffectState()
    state.curiosity   = 0.6
    state.engagement  = 0.9
    state.uncertainty = 0.1
    state.rapport     = 0.8
    state.energy      = 0.7

    via_method = state.to_vad()
    via_factory = AffectVad.from_state(state)

    assert via_method == via_factory
